aws_signing_helper/serve.go

```diff
@@ -0,0 +1,270 @@
+var tokenMap = make(map[string]time.Time)
+
+// Generates a random string with the specified length
+func generateToken(length int) (string, error) {
+	randomBytes := make([]byte, 128)
+	_, err := rand.Read(randomBytes)
+	if err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(randomBytes)[:length], nil
+}
+
+// Removes the token that expires the earliest
+func RemoveEarliestExpiringToken() error {
+	if len(tokenMap) == 0 {
+		return errors.New("unable to remove from an empty map")
+	}
+	earliestExpirationTime := time.Unix(1<<63-1, 0)
+	var earliestExpiringToken string
+	for key, value := range tokenMap {
+		if earliestExpirationTime.After(value) {
+			earliestExpiringToken = key
+			earliestExpirationTime = value
+		}
+	}
+
+	delete(tokenMap, earliestExpiringToken)
+	log.Printf("evicting earliest expiring token: %s", earliestExpiringToken)
+	return nil
+}
+
+// Helper function that checks to see whether the token provided in the request is valid
+func checkValidToken(w http.ResponseWriter, r *http.Request) error {
+	token := r.Header.Get(EC2_METADATA_TOKEN_HEADER)
+	if token == "" {
+		w.WriteHeader(http.StatusUnauthorized)
+		msg := "no token provided"
+		io.WriteString(w, msg)
+		return errors.New(msg)
+	}
+
+	mutex.Lock()
+	expiration, ok := tokenMap[token]
+	mutex.Unlock()
+	if ok {
+		if time.Now().After(expiration) {
+			w.WriteHeader(http.StatusUnauthorized)
+			msg := "invalid token provided"
+			io.WriteString(w, msg)
+			return errors.New(msg)
+		}
+	} else {
+		w.WriteHeader(http.StatusUnauthorized)
+		msg := "invalid token provided"
+		io.WriteString(w, msg)
+		return errors.New(msg)
+	}
+
+	return nil
+}
+
+func AllIssuesHandlers(cred *RefreshableCred, roleName string, opts *CredentialsOpts) (http.HandlerFunc, http.HandlerFunc, http.HandlerFunc) {
+	// Handles PUT requests to /latest/api/token/
+	putTokenHandler := func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != "PUT" {
+			w.WriteHeader(http.StatusMethodNotAllowed)
+			return
+		}
+
+		// Check for the presence of the X-Forwarded-For header
+		xForwardedForHeader := r.Header.Get(X_FORWARDED_FOR_HEADER)
+		if xForwardedForHeader != "" {
+			w.WriteHeader(http.StatusBadRequest)
+			io.WriteString(w, "unable to process requests with X-Forwarded-For header")
+			return
+		}
+
+		// Obtain the token TTL
+		tokenTTLStr := r.Header.Get(EC2_METADATA_TOKEN_TTL_HEADER)
+		if tokenTTLStr == "" {
+			tokenTTLStr = DEFAULT_TOKEN_TTL_SECONDS
+		}
+		tokenTTL, err := strconv.Atoi(tokenTTLStr)
+		if err != nil || tokenTTL < 1 || tokenTTL > 21600 {
+			w.WriteHeader(http.StatusBadRequest)
+			io.WriteString(w, "invalid token TTL")
+			return
+		}
+
+		// Generate token and insert it into map
+		token, err := generateToken(100)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			io.WriteString(w, "unable to generate token")
+			return
+		}
+		expirationTime := time.Now().Add(time.Second * time.Duration(tokenTTL))
+		mutex.Lock()
```

Comment:
If you keep the mutex (rather than using sync.Map), I'd put this entire block inside of `RemoveEarliestExpiringToken` - in general, I'd encourage data structures to be modified right alongside the mutex that controls access.